### PR TITLE
TLS 1.2/1.3 version negotiation - 1

### LINKED ISSUE #4823: Allow runtime choice of TLS 1.2 and TLS 1.3
Adjust TLS 1.3 prototype implementation so that the choice of TLS 1.2 and TLS 1.3 can be negotiated at runtime.

Also, study relation between `mbedtls_ssl_session.minor_ver` and `mbedtls_ssl_context.minor_ver` and whether one of them can be removed. See https://github.com/ARMmbed/mbedtls/pull/4806#discussion_r680510002.

Depends on: https://github.com/ARMmbed/mbedtls/issues/4224

### PR DESCRIPTION
**Pull request by ronald-cron-arm**
## Description
First PR to add support for TLS 1.2/1.3 version negotiation.
This PR does some preparation work, mostly code re-organization and simplification of TLS version related code.

THE FIRST COMMIT IS NOT PART OF THIS PR AND AS TO BE REVIEWED IN #5630.

Fix #4823.

## Status
READY

## Requires Backporting
NO  

## Migrations
NO

## Additional comments
Depends on #5630

### CODE REVIEW

File: library/ssl_tls.c [hunk 1]
```diff
@@ -3089,8 +3122,8 @@ void mbedtls_ssl_handshake_free( mbedtls_ssl_context *ssl )
     mbedtls_ssl_buffering_free( ssl );
 #endif /* MBEDTLS_SSL_PROTO_DTLS */
 
-#if defined(MBEDTLS_ECDH_C) &&                  \
-    defined(MBEDTLS_USE_PSA_CRYPTO)
+#if defined(MBEDTLS_ECDH_C) && \
+    ( defined(MBEDTLS_USE_PSA_CRYPTO) || defined(MBEDTLS_SSL_PROTO_TLS1_3) )
```
Comment (yuhaoth) [inline library/ssl_tls.c:3126]: Both `MBEDTLS_USE_PSA_CRYTO` and `MBEDTLS_SSL_PROTO_TLS1_3` depend on `MBEDTLS_PSA_CRYPTO_C`
```suggestion
#if defined(MBEDTLS_ECDH_C) && defined(MBEDTLS_PSA_CRYPTO_C) 
```

**Review by yuhaoth — CHANGES_REQUESTED** [review]

**Comment by ronald-cron-arm** [comment]
Trivial rebase to fix the conflict in ssl_srv.c.

**Comment by ronald-cron-arm** [comment]
Trivial rebase to fix the conflict in ssl_cli.c.

File: library/ssl_tls.c (hunk 1 above)
Comment (ronald-cron-arm) [reply library/ssl_tls.c:3126]: Yes that would be another possibility but the current version reflects better the current status of the code base regarding usage of PSA in TLS I'd say. Thus not changed.

**Comment by ronald-cron-arm** [comment]
I rebased and with the conflicts due to the merge of #5523. Conflicts occurred only with the last commit.

**Comment by ronald-cron-arm** [comment]
I've rebased to resolve the conflicts resulting of the merge of https://github.com/ARMmbed/mbedtls/pull/5630. Straightforward rebase.

**Review by yuhaoth — DISMISSED** [review]
LGTM

File: library/ssl_tls13_client.c [hunk 2]
```diff
@@ -885,62 +942,49 @@ static int ssl_tls13_write_client_hello_body( mbedtls_ssl_context *ssl,
...
 
-    /* Echo the cookie if the server provided one in its preceding
-     * HelloRetryRequest message.
-     */
-    ret = ssl_tls13_write_cookie_ext( ssl, p, end, &output_len );
-    if( ret != 0 )
-        return( ret );
-    p += output_len;
-
+#if defined(MBEDTLS_SSL_PROTO_TLS1_3)
```
Comment (paul-elliott-arm) [inline library/ssl_tls13_client.c:969]: Is this required? This is after all the tls1.3 client, and this is checked at the top of the file.

**Review by paul-elliott-arm — COMMENTED** [review]
Only one minor comment, otherwise all looks good to me.

**Review by paul-elliott-arm — APPROVED** [review]
LGTM

**Comment by ronald-cron-arm** [comment]
@yuhaoth still ok with you ?

**Review by yuhaoth — APPROVED** [review]
LGTM

**Pull request merged**